### ISSUE TITLE
Bad results from Kmeans

### ISSUE THREAD

**Issue by datvuthanh** [mnslarcher/kmeans-anchors-ratios#14]
Hi mnslarcher,

I implemented your code to my efficiendet code with BDD100K dataset. But I got the result:
```
[01/22 06:17:19] Starting the calculation of the optimal anchors ratios
[01/22 06:17:19] Extracting and preprocessing bounding boxes
[01/22 06:17:21] Discarding 0 bounding boxes with size lower or equal to 0
[01/22 06:17:21] K-Means (10 runs): 100%|███████████████| 10/10 [00:22<00:00,  2.24s/it]
	Runs avg. IoU: 85.10% ± 0.00% (mean ± std. dev. of 10 runs, 0 skipped)
	Avg. IoU between bboxes and their most similar anchors after norm. them to make their area equal (only ratios matter): 85.10%

[01/22 06:17:45] Default anchors ratios: [(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]
	Avg. IoU between bboxes and their most similar default anchors, no norm. (both ratios and sizes matter): 38.23%
	Num. bboxes without similar default anchors (IoU < 0.5):  637020/999085 (63.76%)
[01/22 06:17:48] K-Means anchors ratios: [(0.8, 1.3), (1.1, 0.9), (1.3, 0.8)]
	Avg. IoU between bboxes and their most similar K-Means anchors, no norm. (both ratios and sizes matter): 38.46%
	Num. bboxes without similar K-Means anchors (IoU < 0.5):  643636/999085 (64.42%)
[01/22 06:17:49] Default anchors have an IoU < 50% with bboxes in 0.66% less cases than the K-Means anchors, you should consider stick with them
```

38,46% is a bad result, right? Thank you!

**Comment by mnslarcher** [comment]
Hi @datvuthanh !

Your main problem doesn't seem to be the ratios but the sizes. If you look at the results ignoring the size you get 85% but then when you also consider the size you get more or less 38% (both with the default and optimized ratios which are basically the same in this case).

My guess is that your objects are too small (usually this is the main problem) so you should consider adding a smaller size or accept that you will have problems with very small objects.

Note that this repo only optimizes ratios so sizes you have to test them manually (but it's pretty easy).

I'm not plaing with EfficientDet from a while but for example in the official repo you have to look at these HPs [here](https://github.com/google/automl/blob/d5afa2a63dfcf25ca4b1dff2f2de8146d3266553/efficientdet/hparams_config.py#L207):

```
  h.num_scales = 3
  h.anchor_scale = 4.0
```

Note that you can use this repo to simulate what happens if you change the sizes, have a look at the tutorial and change the anchors_sizes parameter.

**Comment by datvuthanh** [comment]
Hi @mnslarcher,

Thank for your reply, I switched to use EfficientDet for my project and facing many challenges. 

1) The first problem is my network is trained about 150 epochs and reg loss stuck at 0.69 (although I just train only one class) and they don't decrease anymore. So I think, maybe the main problem my network cant detect small objects. This is the reason why I used your repo and recalculated anchor aspects. Do you have any advices for me?

2) How I can add a smaller size (I am not very clear your reply), sorry about that. This means that I need to edited anchor_scale parameter is smaller than 4?

3) I used kmeans code from yolov5 to recalculate my anchors and this is output (https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py):
```
tensor([[[0.5579, 1.4929],
[0.8694, 2.4489],
[1.3794, 3.4020]],
[[0.9983, 2.6222],
[1.7004, 3.8757],
[2.7403, 5.7757]],
[[1.2987, 6.0161],
[2.5212, 5.3684],
[3.3681, 9.0867]]]
```
This is 9 anchor boxes, so how I can tranform these anchors to efficientdet format? 

Thank you!

**Comment by mnslarcher** [comment]
> Hi @mnslarcher,
> 
> Thank for your reply, I switched to use EfficientDet for my project and facing many challenges.
> 
>     1. The first problem is my network is trained about 150 epochs and reg loss stuck at 0.69 (although I just train only one class) and they don't decrease anymore. So I think, maybe the main problem my network cant detect small objects. This is the reason why I used your repo and recalculated anchor aspects. Do you have any advices for me?
> 
>     2. How I can add a smaller size (I am not very clear your reply), sorry about that. This means that I need to edited anchor_scale parameter is smaller than 4?
> 
>     3. I used kmeans code from yolov5 to recalculate my anchors and this is output (https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py):
> 
> 
> ```
> tensor([[[0.5579, 1.4929],
> [0.8694, 2.4489],
> [1.3794, 3.4020]],
> [[0.9983, 2.6222],
> [1.7004, 3.8757],
> [2.7403, 5.7757]],
> [[1.2987, 6.0161],
> [2.5212, 5.3684],
> [3.3681, 9.0867]]]
> ```
> 
> This is 9 anchor boxes, so how I can tranform these anchors to efficientdet format?
> 
> Thank you!

In this repo you can look at the [tutorial.ipynb](https://github.com/mnslarcher/kmeans-anchors-ratios/blob/master/tutorial.ipynb) to understand how to generate the anchors sizes for EfficientDet and simulate what would be the effect of changing the hyper-parameters that I mentioned in my previous answer:

```
## change the following parameters according to your model:

# EfficientDetD{PHI}
PHI = 0  # for another efficientdet change only this, e.g. PHI = 3 for D3

input_sizes = [512, 640, 768, 896, 1024, 1280, 1280, 1536, 1536]
pyramid_levels = [5, 5, 5, 5, 5, 5, 5, 5, 6]
anchor_scale = [4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 4.0]

scale = anchor_scale[PHI]
strides = 2 ** np.arange(3, pyramid_levels[PHI] + 3)
scales = np.array([2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)])

INPUT_SIZE = input_sizes[PHI]
ANCHORS_SIZES = (scale * scales * strides[:, np.newaxis]).flatten().tolist()
ANCHORS_SIZES
```
In your case you want to change this line:
`scale = anchor_scale[PHI]`
with, for example:
`scale = 2.0`

That is like changing the parameter 
` h.anchor_scale = 2.0`

of my previous reply (if you are using the TensorFlow EfficientDet).

This assume that your input_size is the default one (es. 512 if you are using EfficientDet0), otherwise change also that one.

Then you can also increase the number of scales:

This would be the parameter 
`  h.num_scales = 3`

Of my previous reply.

To simulate this is the notebook change, for example:
scales = np.array([2 ** 0, 2 ** (1.0 / 3.0), 2 ** (2.0 / 3.0)])

To 
scales = np.array([2 ** 0, 2 ** (1.0 / 4.0), 2 ** (2.0 / 4.0), 2 ** (3.0 / 4.0)])

That is equivalent to:
`  h.num_scales = 4`

Notice that scales are always like

`2 ** 0, 2 ** (1.0 / num_scales), ..., 2 ** (num_scales - 1 / num_scales)`

Then run:

```
from kmeans_anchors_ratios import get_optimal_anchors_ratios

anchors_ratios = get_optimal_anchors_ratios(
    instances,
    anchors_sizes=ANCHORS_SIZES,
    input_size=INPUT_SIZE,
    normalizes_bboxes=True,
    num_runs=3,
    num_anchors_ratios=3,
    max_iter=300,
    iou_threshold=0.5,
    min_size=0,
    decimals=1,
    default_anchors_ratios=[(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]
)
```
And see what happen, if the situation improve you can use the same parameters for training your model.

**Comment by datvuthanh** [comment]
Hi @mnslarcher,

Thank for your awesome supporting!

But I have a question about when increasing num_scales to 4, the cost computation of network is more expensive? 

This is the output of num_scales = 4 and anchor_scale = 2.0:
```
[01/23 21:27:38] Starting the calculation of the optimal anchors ratios
[01/23 21:27:38] Extracting and preprocessing bounding boxes
[01/23 21:27:39] Discarding 0 bounding boxes with size lower or equal to 0
[01/23 21:27:39] K-Means (10 runs): 100%|███████████████| 10/10 [00:21<00:00,  2.10s/it]
	Runs avg. IoU: 85.10% ± 0.00% (mean ± std. dev. of 10 runs, 0 skipped)
	Avg. IoU between bboxes and their most similar anchors after norm. them to make their area equal (only ratios matter): 85.10%

[01/23 21:28:03] Default anchors ratios: [(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]
	Avg. IoU between bboxes and their most similar default anchors, no norm. (both ratios and sizes matter): 66.86%
	Num. bboxes without similar default anchors (IoU < 0.25):  96587/999085 (9.67%)
[01/23 21:28:06] K-Means anchors ratios: [(0.8, 1.3), (1.1, 0.9), (1.3, 0.8)]
	Avg. IoU between bboxes and their most similar K-Means anchors, no norm. (both ratios and sizes matter): 67.16%
	Num. bboxes without similar K-Means anchors (IoU < 0.25):  98659/999085 (9.87%)
[01/23 21:28:06] Default anchors have an IoU < 25% with bboxes in 0.21% less cases than the K-Means anchors, you should consider stick with them
```

**Comment by mnslarcher** [comment]
> Hi @mnslarcher,
> 
> 
> 
> Thank for your awesome supporting!
> 
> 
> 
> But I have a question about when increasing num_scales to 4, the cost computation of network is more expensive? 

Yes it is, is like using 4 ratios instead of 3, there is a trade off between this and performance, you have to value if it's worthy or not.

As a first experiment I'd just try to lower the scale from 4 to 2 and see what happen, by the way in the tutorial you can also find out what are the bbox that you are missing so to better understand what is your problem (how small they are etc)

**Comment by datvuthanh** [comment]
Hi @mnslarcher,

I edited my reply and you can see it again. Today, I have a experiment with default aspect ratios and anchor scales. ([(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]). I visualized my anchor boxes with ground truth labels and I see that my anchor boxes with default settings can be detect small objects. 

So, I have a question? What should I do now? Keep training with default settings or replacing by your approach? Thank you!

**Comment by datvuthanh** [comment]
In my experiment, I set up hyparameters (default settings):

```
anchors_scales: '[2 ** 0, 2 ** 1/3, 2 ** 2/3]'
anchors_ratios: '[(0.7, 1.4), (1.0, 1.0), (1.4, 0.7)]'
```

And I used Cli98 repos to visualize my anchor boxes. This is the output, IoU between AnchorBoxes and Ground Truth Labels (Iou Threshold = 0.5):
![269776709_652361542629343_4934753148303585680_n](https://user-images.githubusercontent.com/44630913/150699583-7541db40-7e9c-4ac8-8431-234b459d2951.png)
![269776709_652361542629343_4934753148303585680_n](https://user-images.githubusercontent.com/44630913/150699599-574c9617-27d8-41ca-91e2-128d2fc2dfe6.png)
![271758547_647557173032388_8480832336091576039_n](https://user-images.githubusercontent.com/44630913/150699594-72f4f27d-be33-4eef-a791-6f03b6e76afc.png)

And Iou Threshold = 0.25:

![271816269_620129759219760_6477922120037210463_n](https://user-images.githubusercontent.com/44630913/150699622-176647e3-c051-46ac-bfde-20bbf982363d.png)
![271283082_1563142820731575_3623866376459228957_n](https://user-images.githubusercontent.com/44630913/150699631-e8d39351-e3ce-4c9f-a21e-51e96597af9c.png)
![271831455_1251104288723068_1004471101575016491_n](https://user-images.githubusercontent.com/44630913/150699628-9e2f6773-bebc-4037-98c4-b8993dce6596.png)

The red boxes which my anchor boxes missed (The green boxes have IoU greater than IoU threshold). But I believe that my regression network can be fit that boxes. Do you agree that? 

Please give some advices for me in this situation. 

Thank you a lots!

**Comment by mnslarcher** [comment]
Hi @datvuthanh, you already know that with that setting (the default setting) you will have 64.42% of anchors with less than 50% IoU with the GT and so you will risk to miss many objects or to make the life hard for your model.

I don't know how much the results improve if, for example, you set anchor_scale = 2.0 because you posted the result with IoU < 25% (that is not a fair comparison with the previously posted result were you used correctly 50% < IoU).

This repo optimize the ratios but in the case of you dateset the default or optimal ratios are basically the same (you can also leave them as they are). What you can try to change is the sizes but again to understand what to do you have to compare apples with apples (do the same experiment without changing the IoU threshold).

Another thing that can improve you results a little is to use 4 ratios instead of 3 but this increase the computation (changing the scale to 2.0 doesn't do it).

Then in the end you can always train one time your model in one way and one time in the other way and see the results so you will be able to understand better the full effect of changing that parameter

**Comment by datvuthanh** [comment]
Hi @mnslarcher,

Thank you for your supporting very much!

Firstly, I trained with default settings about 50 epochs and after 20th epoch my regression loss cannot decrease (although I believe that if keeping training reg loss can decrease). I stopped training to transfer your approach.

Secondly, I changed scale 2.0 and based on autoanchor of yolov5 to find my aspect ratio and scale:

```
anchors_scales: '[2 ** 0, 2 ** 0.70, 2 ** 1.32]'
anchors_ratios: '[(0.62, 1.58), (1.0, 1.0), (1.58, 0.62)]'
```

And there are my results on IoU 0.5 (some anchor boxes can detect small objects):

![272149633_607386053667300_385980133087995363_n](https://user-images.githubusercontent.com/44630913/150876299-da056c26-0973-42cb-9339-7faa4870d63e.png)
![271930486_695503151813279_2794803711134392729_n](https://user-images.githubusercontent.com/44630913/150876314-cf11e60e-fbdb-4b73-b12f-e6c1a5dc6d46.png)
![271807032_4702964213152887_4366557909817609426_n](https://user-images.githubusercontent.com/44630913/150876320-790899b0-1b97-4172-a372-fb8833dc1078.png)
![271837597_322357869840691_3000966899475178865_n](https://user-images.githubusercontent.com/44630913/150876321-ddfbd2b5-edb0-41d8-89ff-cf61d3af5301.png)
![271784060_472668134387866_2707735349038955016_n](https://user-images.githubusercontent.com/44630913/150876348-f6d71540-4cdc-4001-bf67-d527cf33c844.png)


I feel that everything is very clear at now, thank you again!

**Comment by mnslarcher** [comment]
You are welcome @datvuthanh, I close the issue
